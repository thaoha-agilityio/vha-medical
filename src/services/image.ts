'use server';

import { IMGBB_URL } from '@/constants';
import { ImgBBResponse } from '@/types';

export const uploadImageToImgbb = async (payload: FormData) => {
  try {
    const response = await fetch(IMGBB_URL, {
      headers: undefined,
      body: payload,
      method: 'POST',
    });

    if (!response.ok) {
      const error = await response.text();

      throw new Error(error);
    }

    const {
      data: { url },
    }: ImgBBResponse = await response.json();
    return { image: url, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : 'An unexpected error occurred when upload image';

    return { image: null, error: errorMessage };
  }
};
