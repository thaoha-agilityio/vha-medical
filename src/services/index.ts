export * from './api';
export * from './appointment';
export * from './auth';
export * from './chemists';
export * from './image';
export * from './notification';
export * from './specialty';
export * from './user';
