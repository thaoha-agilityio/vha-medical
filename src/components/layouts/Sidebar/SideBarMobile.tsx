'use client';

import {
  cn,
  Modal,
  ModalBody,
  ModalContent,
  useDisclosure,
} from '@nextui-org/react';

// Services
import { logout } from '@/services';

// Components
import { Button, Image, Navbar } from '@/components/ui';
import { ArrowRightIcon, LogoutIcon } from '@/icons';

// Constants
import { AUTH_ROUTES, PRIVATE_ROUTES, SRC_LOGO } from '@/constants';
import Link from 'next/link';

export const SidebarMobile = () => {
  const { isOpen, onOpen, onOpenChange, onClose } = useDisclosure();

  const onLogout = async () => {
    await logout();
    location.replace(AUTH_ROUTES.LOGIN);
  };

  return (
    <aside>
      <div className="flex absolute left-0">
        <Button
          className="p-0 min-w-4 h-7 bg-linear-success rounded-none rounded-r-lg lg:hidden"
          onClick={onOpen}
          data-testid="open-sidebar-mobile"
          aria-label="open sidebar"
        >
          <ArrowRightIcon customClass="w-4 h-4" />
        </Button>
      </div>
      <Modal
        isOpen={isOpen}
        onOpenChange={onOpenChange}
        className={cn(
          {
            'animate-slideInLeft': isOpen,
            'animate-slideInRight': !isOpen,
          },
          'm-0 sm:m-0 rounded-none max-w-[277px] lg:hidden',
        )}
        classNames={{
          base: 'absolute top-0 left-0',
          closeButton:
            'p-0 min-w-4 h-7 bg-linear-success rounded-none rounded-l-lg text-content1 top-2 right-0',
          backdrop: 'lg:hidden z-[40]',
        }}
        closeButton={
          <Button onClick={onClose} aria-label="close sidebar">
            <ArrowRightIcon customClass="w-4 h-4 rotate-180" />
          </Button>
        }
      >
        <ModalContent>
          <ModalBody className="min-h-screen p-0 bg-background-200">
            <h1 className="m-auto py-5">
              <Link href={PRIVATE_ROUTES.DASHBOARD}>
                <Image
                  src={SRC_LOGO}
                  alt="logo"
                  width={40}
                  height={40}
                  placeholder="empty"
                  className="w-full h-auto"
                />
              </Link>
            </h1>
            <div className="flex-1">
              <Navbar isExpandSidebar />
            </div>
            <div className="pl-3 py-5">
              <Button
                color="stone"
                startContent={<LogoutIcon />}
                className="gap-3"
                onClick={onLogout}
              >
                Logout
              </Button>
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>
    </aside>
  );
};
