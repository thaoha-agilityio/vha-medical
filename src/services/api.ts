import { auth } from '@/config/auth';
import { API_URL, AUTH_TOKEN } from '@/constants';

interface IApiClient {
  baseURL: string;
  headers?: HeadersInit;
}

interface RequestInitExtended extends Omit<RequestInit, 'body'> {
  body?: object | null;
}

export class ApiClient {
  baseURL: string;
  config: RequestInit;
  private static apiClientInstance: ApiClient;

  private constructor(baseURL: string, config: RequestInit) {
    this.baseURL = baseURL;
    this.config = config;
  }

  static create(params: IApiClient): ApiClient {
    const { baseURL, headers = { Authorization: `Bearer ${AUTH_TOKEN}` } } =
      params;

    if (!this.apiClientInstance)
      this.apiClientInstance = new ApiClient(baseURL, { headers });

    return this.apiClientInstance;
  }

  async apiClientSession(config: RequestInit = {}) {
    const session = await auth();
    if (!session?.user) {
      throw new Error('Not found user!');
    }

    const { token } = session.user;

    return new ApiClient(API_URL, {
      ...this.config,
      ...config,
      headers: {
        Authorization: `Bearer ${token}`,
        ...config.headers,
      },
    });
  }

  async get<T>(url: string, config: RequestInit = {}): Promise<T> {
    const response = await this.fetchWithConfig(url, config);
    if (!response.ok) {
      const errorText = await response.text();

      return {
        error: errorText,
      } as T;
    }
    const data: T = await response.json();
    return data;
  }

  async post<T>(url: string, config: RequestInitExtended = {}): Promise<T> {
    const response = await this.fetchWithConfig(url, {
      ...config,
      method: 'POST',
      body: JSON.stringify(config.body),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const errorText = await response.text();

      return {
        error: errorText,
      } as T;
    }

    const data: T = await response.json();
    return data;
  }

  async postFile<T>(url: string, config: RequestInit = {}): Promise<T> {
    const response = await this.fetchWithConfig(url, {
      ...config,
      method: 'POST',
      body: config.body,
    });

    if (!response.ok) {
      const errorText = await response.text();

      return {
        error: errorText,
      } as T;
    }

    const data: T = await response.json();
    return data;
  }

  async put<T>(url: string, config: RequestInitExtended = {}): Promise<T> {
    const response = await this.fetchWithConfig(url, {
      ...config,
      method: 'PUT',
      body: JSON.stringify(config.body),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const errorText = await response.text();

      return {
        error: errorText,
      } as T;
    }

    const data: T = await response.json();
    return data;
  }

  async delete<T>(url: string, config?: RequestInit): Promise<T> {
    const response = await this.fetchWithConfig(url, {
      ...config,
      method: 'DELETE',
    });

    if (!response.ok) {
      const errorText = await response.text();

      return {
        error: errorText,
      } as T;
    }

    const data: T = await response.json();
    return data;
  }

  fetchWithConfig(url: string, config?: RequestInit) {
    return fetch(`${this.baseURL}${url}`, {
      ...this.config,
      ...config,
      headers: {
        ...this.config.headers,
        ...config?.headers,
      },
    });
  }
}

export const apiClient = ApiClient.create({
  baseURL: API_URL,
});
