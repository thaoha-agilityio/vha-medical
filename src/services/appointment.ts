import { revalidateTag } from 'next/cache';

// Constants
import { API_ENDPOINT, EXCEPTION_ERROR_MESSAGE } from '@/constants';

// Types
import {
  AppointmentsResponse,
  AppointmentsDataResponse,
  FetchDataProps,
  AppointmentPayload,
  AppointmentResponse,
  ErrorResponse,
  AppointmentDataResponse,
} from '@/types';

// Services
import { apiClient } from '@/services';

export const getAppointments = async ({
  searchParams = new URLSearchParams(),
  options = { next: { tags: [API_ENDPOINT.APPOINTMENTS] } },
}: FetchDataProps): AppointmentsDataResponse => {
  try {
    const params = new URLSearchParams(searchParams);
    const api = await apiClient.apiClientSession();
    const url = decodeURIComponent(
      `${API_ENDPOINT.APPOINTMENTS}?${params.toString()}`,
    );
    const { data, meta, error } = await api.get<
      AppointmentsResponse & { error?: string }
    >(url, {
      ...options,
      next: {
        ...options.next,
        revalidate: 3600,
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { appointments: [], error: errorResponse.error.message };
    }

    return {
      appointments: data,
      ...meta,
      error: null,
    };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.GET('appointments');

    return { appointments: [], error: errorMessage };
  }
};

export const addAppointment = async (
  appointment: AppointmentPayload,
): Promise<AppointmentDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.post<{
      data: AppointmentResponse;
      error?: string;
    }>(`${API_ENDPOINT.APPOINTMENTS}`, {
      body: {
        data: appointment,
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { appointment: null, error: errorResponse.error.message };
    }

    revalidateTag(`${API_ENDPOINT.APPOINTMENTS}/dashboard`);
    revalidateTag(API_ENDPOINT.APPOINTMENTS);

    return { appointment: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.ADD('appointment');

    return { appointment: null, error: errorMessage };
  }
};

export const updateAppointment = async (
  id: string,
  appointment: AppointmentPayload,
): Promise<AppointmentDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.put<{
      data: AppointmentResponse;
      error?: string;
    }>(`${API_ENDPOINT.APPOINTMENTS}/${id}`, {
      body: {
        data: {
          ...appointment,
        },
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { appointment: null, error: errorResponse.error.message };
    }

    revalidateTag(`${API_ENDPOINT.APPOINTMENTS}/dashboard`);
    revalidateTag(API_ENDPOINT.APPOINTMENTS);

    return { appointment: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.UPDATE('appointment');

    return { appointment: null, error: errorMessage };
  }
};

export const deleteAppointment = async (
  id: string,
): Promise<AppointmentDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.delete<{
      data: AppointmentResponse;
      error?: string;
    }>(`${API_ENDPOINT.APPOINTMENTS}/${id}`);

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { appointment: null, error: errorResponse.error.message };
    }

    revalidateTag(`${API_ENDPOINT.APPOINTMENTS}/dashboard`);
    revalidateTag(API_ENDPOINT.APPOINTMENTS);

    return { appointment: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.DELETE('appointment');

    return { appointment: null, error: errorMessage };
  }
};
