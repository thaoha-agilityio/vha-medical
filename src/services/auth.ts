'use server';

import { apiClient } from './api';
import { getUserLogged } from './user';
import {
  UserSession,
  AuthResponse,
  LoginFormData,
  SignupFormData,
  ErrorResponse,
} from '@/types';
import {
  API_ENDPOINT,
  AVATAR_THUMBNAIL,
  EXCEPTION_ERROR_MESSAGE,
} from '@/constants';
import { signOut } from '@/config/auth';
import { cookies } from 'next/headers';
import { unregisterFCM } from './notificationFirebase';
import { decrypt } from '@/utils/encode';

export const login = async (
  body: LoginFormData,
): Promise<{ user: UserSession | null; error: string | null }> => {
  try {
    const response = await apiClient.post<AuthResponse>(API_ENDPOINT.AUTH, {
      body: {
        identifier: body.identifier,
        password: body.password,
      },
    });

    const { error, jwt, user } = response;

    if (error && !user) {
      return {
        user: null,
        error: (JSON.parse(error) as ErrorResponse).error.message,
      };
    }
    const { user: userLogged, error: errorGetUserLogged } =
      await getUserLogged(jwt);

    if (errorGetUserLogged) {
      return {
        user: null,
        error: errorGetUserLogged,
      };
    }
    const {
      id = '',
      role,
      username = '',
      email = '',
      avatar = AVATAR_THUMBNAIL,
    } = userLogged || {};
    const { name = '' } = role || {};

    const data = {
      id: id,
      token: jwt,
      remember: body.remember,
      role: name,
      avatar,
      username,
      email,
    };

    return { user: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error ? error.message : EXCEPTION_ERROR_MESSAGE.LOGIN;

    return {
      user: null,
      error: errorMessage,
    };
  }
};

export const signup = async (
  body: Omit<SignupFormData, 'confirmPassWord'>,
): Promise<AuthResponse> => {
  try {
    const {
      error,
      user,
      jwt = '',
    } = await apiClient.post<AuthResponse>(`${API_ENDPOINT.AUTH}/register`, {
      body,
    });

    if (error && !user) {
      return {
        user: null,
        error: (JSON.parse(error) as ErrorResponse).error.message,
        jwt,
      };
    }

    return { user, error: null, jwt };
  } catch (error) {
    const errorMessage =
      error instanceof Error ? error.message : EXCEPTION_ERROR_MESSAGE.REGISTER;

    return {
      user: null,
      error: errorMessage,
      jwt: '',
    };
  }
};

export const logout = async () => {
  const encryptedValue = cookies().get('fcm_token')?.value;

  if (encryptedValue) {
    const token = await decrypt(encryptedValue);

    if (token) {
      await unregisterFCM({ token });

      cookies().delete('fcm_token');
    }
  }

  await signOut();
};
