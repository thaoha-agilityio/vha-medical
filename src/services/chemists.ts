'use server';
import { revalidateTag } from 'next/cache';

// Types
import {
  ChemistDataResponse,
  ChemistPayload,
  ChemistResponse,
  ChemistsDataResponse,
  ChemistsResponse,
  ErrorResponse,
  FetchDataProps,
} from '@/types';

// Services
import { apiClient } from './api';

// Constants
import { API_ENDPOINT, EXCEPTION_ERROR_MESSAGE } from '@/constants';

export const addUserToChemists = async (
  payload: ChemistPayload,
): Promise<ChemistDataResponse> => {
  try {
    const { data, error } = await apiClient.post<{
      data: ChemistResponse;
      error?: string;
    }>(`${API_ENDPOINT.CHEMISTS}`, {
      body: {
        data: payload,
      },
    });

    if (error) {
      return {
        chemist: null,
        error: (JSON.parse(error) as ErrorResponse).error.message,
      };
    }

    revalidateTag(API_ENDPOINT.CHEMISTS);

    return { chemist: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.ADD('user to chemists');

    return { chemist: null, error: errorMessage };
  }
};

export const getChemists = async ({
  searchParams = new URLSearchParams(),
  options = { next: { tags: [API_ENDPOINT.CHEMISTS] } },
}: FetchDataProps): ChemistsDataResponse => {
  try {
    const api = await apiClient.apiClientSession();
    const url = decodeURIComponent(
      `${API_ENDPOINT.CHEMISTS}?${searchParams.toString()}`,
    );

    const { data, meta, error } = await api.get<
      ChemistsResponse & { error?: string }
    >(url, {
      ...options,
      next: {
        ...options.next,
        revalidate: 3600,
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { chemists: [], error: errorResponse.error.message };
    }

    return {
      chemists: data,
      ...meta,
      error: null,
    };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.GET('chemists');

    return { chemists: [], error: errorMessage };
  }
};
