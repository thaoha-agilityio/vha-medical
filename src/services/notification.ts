import { revalidateTag } from 'next/cache';

// Constants
import { API_ENDPOINT, EXCEPTION_ERROR_MESSAGE } from '@/constants';

// Types
import {
  NotificationsResponse,
  NotificationsDataResponse,
  FetchDataProps,
  NotificationPayload,
  NotificationResponse,
  ErrorResponse,
  NotificationDataResponse,
} from '@/types';

// Services
import { apiClient } from '@/services';

export const getNotifications = async ({
  searchParams = new URLSearchParams(),
  options = { next: { tags: [API_ENDPOINT.NOTIFICATIONS] } },
}: FetchDataProps): NotificationsDataResponse => {
  try {
    const params = new URLSearchParams(searchParams);
    const api = await apiClient.apiClientSession();
    const url = decodeURIComponent(
      `${API_ENDPOINT.NOTIFICATIONS}?${params.toString()}`,
    );
    const { data, meta, error } = await api.get<
      NotificationsResponse & { error?: string }
    >(url, {
      ...options,
      next: {
        ...options.next,
        revalidate: 3600,
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { notifications: [], error: errorResponse.error.message };
    }

    return {
      notifications: data,
      ...meta,
      error: null,
    };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.GET('notifications');

    return { notifications: [], error: errorMessage };
  }
};

export const addNotification = async (
  notification: NotificationPayload,
): Promise<NotificationDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.post<{
      data: NotificationResponse;
      error?: string;
    }>(`${API_ENDPOINT.NOTIFICATIONS}`, {
      body: {
        data: notification,
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { notification: null, error: errorResponse.error.message };
    }

    revalidateTag(API_ENDPOINT.NOTIFICATIONS);

    return { notification: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.ADD('notification');

    return { notification: null, error: errorMessage };
  }
};

export const updateNotification = async (
  id: string,
  notification: NotificationPayload,
): Promise<NotificationDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.put<{
      data: NotificationResponse;
      error?: string;
    }>(`${API_ENDPOINT.NOTIFICATIONS}/${id}`, {
      body: {
        data: {
          ...notification,
        },
      },
    });

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { notification: null, error: errorResponse.error.message };
    }

    revalidateTag(API_ENDPOINT.NOTIFICATIONS);

    return { notification: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.UPDATE('notification');

    return { notification: null, error: errorMessage };
  }
};

export const deleteNotification = async (
  id: string,
): Promise<NotificationDataResponse> => {
  try {
    const api = await apiClient.apiClientSession();
    const { data, error } = await api.delete<{
      data: NotificationResponse;
      error?: string;
    }>(`${API_ENDPOINT.NOTIFICATIONS}/${id}`);

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { notification: null, error: errorResponse.error.message };
    }

    revalidateTag(API_ENDPOINT.APPOINTMENTS);

    return { notification: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.DELETE('notification');

    return { notification: null, error: errorMessage };
  }
};
