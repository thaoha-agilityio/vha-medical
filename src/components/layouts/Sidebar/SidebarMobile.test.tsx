import { fireEvent, render, screen } from '@testing-library/react';

// Components
import { SidebarMobile } from './SideBarMobile';

// Services
import { logout } from '@/services';

jest.mock('@/services/auth.ts', () => ({
  logout: jest.fn(),
}));
describe('SidebarMobile test cases', () => {
  const mockLogout = logout as jest.Mock;

  const setup = () => render(<SidebarMobile />);

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should able to logout when clicking logout button', () => {
    setup();

    const openSidebarBtn = screen.getByTestId('open-sidebar-mobile');

    fireEvent.click(openSidebarBtn);

    const logoutBtn = screen.getByRole('button', {
      name: /logout/i,
    });

    fireEvent.click(logoutBtn);

    expect(mockLogout).toHaveBeenCalledTimes(1);
  });
});
