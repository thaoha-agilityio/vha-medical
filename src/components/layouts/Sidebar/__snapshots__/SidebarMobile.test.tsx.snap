// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarMobile test cases should render correctly 1`] = `
<DocumentFragment>
  <aside>
    <div
      class="flex absolute left-0"
    >
      <button
        aria-label="open sidebar"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent font-semibold text-content1 p-0 min-w-4 h-7 bg-linear-success rounded-none rounded-r-lg lg:hidden"
        data-testid="open-sidebar-mobile"
        type="button"
      >
        <svg
          class="w-4 h-4"
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5L15.75 12L8.25 19.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </button>
    </div>
  </aside>
</DocumentFragment>
`;
